### ISSUE TITLE
jep is incompatible with Python 3.9: daemon threads can no longer be used in subinterpreters

### ISSUE THREAD

**Issue by vstinner** [ninia/jep#229]
Hi,

test_shared_modules_threads() fails on Python 3.9 with:

```
FAIL: test_shared_modules_threads (test_shared_modules.TestSharedModules)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/jep-3.9.0/src/test/python/test_shared_modules.py", line 15, in test_shared_modules_threads
    jep_pipe(build_java_process_cmd('jep.test.TestSharedModulesThreads'))
  File "/usr/lib64/python3.9/contextlib.py", line 240, in helper
    return _GeneratorContextManager(func, args, kwds)
  File "/usr/lib64/python3.9/contextlib.py", line 83, in __init__
    self.gen = func(*args, **kwds)
  File "/builddir/build/BUILD/jep-3.9.0/src/test/python/jep_pipe.py", line 36, in jep_pipe
    assert False, stderr
AssertionError: b'Exception in thread "main" jep.JepException: <class \'RuntimeError\'>: daemon thread are not supported in subinterpreters\n\tat /usr/lib64/python3.9/threading.start(threading.py:858)\n\tat <string>.<module>(<string>:1)\n\tat jep.Jep.eval(Native Method)\n\tat jep.Jep.eval(Jep.java:451)\n\tat jep.test.TestSharedModulesThreads.main(TestSharedModulesThreads.java:53)\n'
```

I'm the one who modified Python to deny spawning a daemon thread in a Python subinterpreter: https://bugs.python.org/issue37266 See the issue for the rationale.

Would it be possible to modify jep to use regular threads rather than daemon threads in subinterpreters?

**Comment by bsteffensmeier** [comment]
I don't think Jep actually needs to use daemon threads, it looks like it is just the setup for that one test that is broken. We can change that test to no longer use daemon threads and the test will still be valid. Since Jep allows arbitrary execution of python it is very possible other projects using Jep could also be affected by the Python change.

**Comment by vstinner** [comment]
I forgot to mention the Fedora downstream issue: https://bugzilla.redhat.com/show_bug.cgi?id=1792062

**Comment by bsteffensmeier** [comment]
This is fixed in the dev_3.9 branch with a31d461 and will be part of the 3.9.1 release later this year.

**Comment by bsteffensmeier** [comment]
We no longer have problems with daemon threads, however there are new problems with Python 3.9b1. The error can be seen in the [travis logs](https://travis-ci.org/github/ninia/jep/jobs/690217405), and I have repeated it locally. I have tested Python 3.9a6 and Jep works fine so it is a recent change in Python that has broken things. It tried to track when things brokes and it stopped working after Python change [0b1e3307](https://github.com/python/cpython/commit/0b1e3307e24b0af45787ab6456535b8346e0239a). Jep was working fine in the two versions before that but started having crashes in `PyEval_ReleaseLock()` after that change.

The process crashes whenever we call `PyEval_ReleaseLock()` after closing an interpreter. `PyEval_ReleaseLock()` has been deprecated, but I don't see any other API we could possibly use to free locks after closing an interpreter. The problems with that deprecation are described well in [this unanswered stackoverflow question](https://stackoverflow.com/questions/57994737/calling-py-endinterpreter-from-a-c-worker-thread).

@vstinner  Can you comment on whether our problems are a Python bug that could be fixed before the final 3.9 release or if this is something we need to change Jep to workaround?

**Comment by vstinner** [comment]
> The process crashes whenever we call PyEval_ReleaseLock() after closing an interpreter. 

That's wrong: you must not call PyEval_ReleaseLock() after closing an interpreter.

> I don't see any other API we could possibly use to free locks after closing an interpreter

PyEval_ReleaseLock() releases the GIL, it doesn't free memory. FYI the GIL object is kept alive after Py_Finalize() on purpose, to support daemon threads. It's only destroyed when you exit Py_RunMain() or when Py_Initialize() is called again. In short, it's not possible to explicitly destroy the GIL.

I'm not sure which kind of issue you are trying to solve. An interpreter contains Python thread states. When an interpreter is destroyed, all Python thread states are destroyed as well.

**Comment by bsteffensmeier** [comment]
@vstinner Thank you for looking at this, I got a little loose with some terminology so please let me try to clarify.

> > The process crashes whenever we call PyEval_ReleaseLock() after closing an interpreter.
> 
> That's wrong: you must not call PyEval_ReleaseLock() after closing an interpreter.
> 

According to the [Python Docs](https://docs.python.org/3.9/c-api/init.html#c.Py_EndInterpreter) when we end a sub interpreter with Py_EndInterpreter we still hold the GIL. For other python threads to run we must release the GIL so we must call PyEval_ReleaseLock().

> > I don't see any other API we could possibly use to free locks after closing an interpreter
> 
> PyEval_ReleaseLock() releases the GIL, it doesn't free memory. FYI the GIL object is kept alive after Py_Finalize() on purpose, to support daemon threads. It's only destroyed when you exit Py_RunMain() or when Py_Initialize() is called again. In short, it's not possible to explicitly destroy the GIL.

I meant we must **release** the lock, _free_ was a poor choice of words. We never destroy the main interpreter so we are not concerned with actual freeing the GIL.

> I'm not sure which kind of issue you are trying to solve. An interpreter contains Python thread states. When an interpreter is destroyed, all Python thread states are destroyed as well.

We create and destroy sub-interpreters. We also create and destroy thread states on a  "main" interpreter.[ The code we use for destruction](https://github.com/ninia/jep/blob/dev_3.9/src/main/c/Jep/pyembed.c#L828) was working up until 3.9a6 but will crash in 3.9b1 when we call PyEval_ReleaseLock(). There are some changes we could make to handle deletion of thread states differently to avoid calling PyEval_RelaseLock() but I cannot find any alternative API for ending a sub-interpreter. I suspect this is a regression in Python and I was hoping you could either confirm that or point out the correct API for ending a sub-interpreter and releasing the GIL.

**Comment by bsteffensmeier** [comment]
I've written a small, isolated C program to demonstrate our problem. It includes three different versions of code for managing the GIL while creating a sub-interpreters that work in different versions of Python.

```c
#include "Python.h"
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <pthread.h>

static PyThreadState* main_state;

/*
 * This is how it is done in jep 3.3
 * Works in python 2.7
 * Broke in python 3.x, before python 3.5
 * ndjensen changed jep in 5d4b51ec
 */
void *on_thread_1(void *vargp){
    PyEval_AcquireLock();
    PyThreadState* state = Py_NewInterpreter();
    PyRun_SimpleString("print('In a sub interpreter')\n");
    Py_EndInterpreter(state);
    PyEval_ReleaseLock();
    return NULL; 
} 
   
/*
 * This is how it is done in jep 3.4 - 3.9.0
 * Works in every python version through 3.9a6
 * Broke in python 3.9b1 after 0b1e3307
 */
void *on_thread_2(void *vargp){
    PyEval_AcquireThread(main_state);
    PyThreadState* state = Py_NewInterpreter();
    PyRun_SimpleString("print('In a sub interpreter')\n");
    Py_EndInterpreter(state);
    PyEval_ReleaseLock();
    return NULL; 
} 

/*
 * Is this how we must do it in python 3.9?
 */
void *on_thread_3(void *vargp){
    PyEval_AcquireThread(main_state);
    PyThreadState* state = Py_NewInterpreter();
    PyRun_SimpleString("print('In a sub interpreter')\n");
    Py_EndInterpreter(state);
    PyThreadState_Swap(main_state);
    PyEval_ReleaseThread(main_state);
    return NULL;
}

int main() {
    Py_Initialize();
    PyEval_InitThreads();
    main_state = PyThreadState_Get();
    PyRun_SimpleString("print('Before creating sub interpreter')\n");
    Py_BEGIN_ALLOW_THREADS

    pthread_t thread_id;
    // pthread_create(&thread_id, NULL, on_thread_1, NULL);
    // pthread_create(&thread_id, NULL, on_thread_2, NULL);
    pthread_create(&thread_id, NULL, on_thread_3, NULL);
    pthread_join(thread_id, NULL);

    Py_END_ALLOW_THREADS
    PyRun_SimpleString("print('After running sub interpreter')\n");
    Py_Finalize();

    exit(0);
}
```

**Comment by vstinner** [comment]
IHMO pyembed_thread_close() should call PyThreadState_DeleteCurrent() rather than calling PyEval_ReleaseLock() directly.

Try to replace:

```
    if (jepThread->tstate->interp == mainThreadState->interp) {
        PyThreadState_Clear(jepThread->tstate);
        PyThreadState_Swap(NULL);
        PyThreadState_Delete(jepThread->tstate);
    } else {
        Py_EndInterpreter(jepThread->tstate);
    }
    free(jepThread);
    PyEval_ReleaseLock();
```

with (not tested):

```
    PyThreadState_Clear(jepThread->tstate);
    PyThreadState_Swap(NULL);
    PyThreadState_Delete(jepThread->tstate);
    if (jepThread->tstate->interp != mainThreadState->interp) {
        Py_EndInterpreter(jepThread->tstate);
    }
    free(jepThread);
```

**Comment by bsteffensmeier** [comment]
@vstinner Thanks for looking into this more. I tested you proposed solution with cpython v3.9b3 and 3.6 and several tests crash with `Fatal Python error: Py_EndInterpreter: thread is not current`

The documentation for Py_EndInterpreter indicate that the given thread state must be current which would not be the case if the thread state was deleted. I think PyThreadState_DeleteCurrent() will actually free the memory for the thread state which would means the pointer we are passing into Py_EndInterpreter is no longer valid.

**Comment by bsteffensmeier** [comment]
I have checked in 83b7e65 on the dev_3.9 branch. We will no longer use PyEval_ReleaseLock since it crashes in Python 3.9. We will need to release Jep 3.9.1 as soon as Python 3.9 is released since Python 3.9 will not be compatible with any existing versions of Jep.

**Comment by rapgro** [comment]
Upcoming Fedora 33 will default to Python 3.9, so this is an important feature. Fedora 32 has Python 3.8 and 31 is with 3.7 as defaults.

**Comment by bsteffensmeier** [comment]
Jep 3.9.1 has changes to work with the changes to the cpython API so this can be closed now that 3.9.1 has been released.

**Comment by vstinner** [comment]
Cool! Thanks for fixing that.
